### PR TITLE
Add Dependency Analysis Action and Dockerfile

### PR DESCRIPTION
**Pull request by naveensrinivasan**
New scorecard action https://github.com/ossf/scorecard-action/issues/1070

- Add workflow to publish dependency analysis Docker image
- Add a new filter function to filter slices
- Add a GetScorecardChecks function to get scorecard checks
- Add a GetScore function to get score of a repo
- Add a Validate function to validate token, owner, repo, commitSHA, and PR
- Add a new action file for OSSF Scorecard dependency analysis

https://github.com/ossf-tests/vulpy/pull/2 - python
https://github.com/ossf-tests/vulpy/pull/1 - GitHub action
https://github.com/naveensrinivasan/scorecard-action/pull/23 - golang

Example of the GitHub action https://github.com/ossf-tests/vulpy/blob/master/.github/workflows/scorecard-dependency.yml

### CODE REVIEW

File: dependency-analysis/main.go
Comment (spencerschrock) [inline dependency-analysis/main.go:95]: having a vulnerability will skip all the other checks for a repository? I know having a vuln will cause the action to fail, but is it not worth having the other output?

File: dependency-analysis/main.go
Comment (spencerschrock) [inline dependency-analysis/main.go:109]: this should be moved either to `Validate` or at least before the API calls

File: Dockerfile-dependency-analysis
Comment (raghavkaul) [inline Dockerfile-dependency-analysis:1]: Nit: Rename to Dependency-analysis.dockerfile so that syntax highlighting works

File: dependency-analysis/README.md
Comment (raghavkaul) [inline dependency-analysis/README.md:14]: Does it make sense to allow users to specify skipped dependencies? Totally fine if that feature is not in this PR.

File: dependency-analysis/README.md
Comment (raghavkaul) [inline dependency-analysis/README.md:57]: Nit: github.com/ needed?

File: dependency-analysis/main.go
Comment (raghavkaul) [inline dependency-analysis/main.go:276]: Can the GitHub client from `go-github` be used?

File: dependency-analysis/main.go
Comment (raghavkaul) [inline dependency-analysis/main.go:74]: Should this be a URL + version? Or is there any other way we can make it clear we're running on Git HEAD rather than the version of a package used.

File: dependency-analysis/main.go
Comment (raghavkaul) [inline dependency-analysis/main.go:75]: Should `err != nil` be handled separately?

File: dependency-analysis/main.go
Comment (raghavkaul) [inline dependency-analysis/main.go:97]: Does `GetScore` on L74 run all checks? If it does, it seems like we throw away some of its work. Maybe we should filter these checks earlier

File: dependency-analysis/main.go
Comment (raghavkaul) [inline dependency-analysis/main.go:147]: Nit: Shouldn't ignore lint, would be easier to create a new error type here.

File: dependency-analysis/main.go
Comment (raghavkaul) [inline dependency-analysis/main.go:118]: ghUser is `GITHUB_ACTOR` which is the user who triggered the workflow - what do we expect GITHUB_ACTOR to be usually? if the workflow is re-triggered manually, ghUser would be the identity of a human repo maintainer - and then we'd try to update their first comment, it seems?

File: dependency-analysis/main.go
Comment (raghavkaul) [inline dependency-analysis/main.go:235]: Nit: Should create an error type.

File: dependency-analysis/main.go
Comment (raghavkaul) [reply dependency-analysis/main.go:305]: Agree that it should be parameterized for testability; also wonder if filename is the right approach? If the user is configuring the DependencyDiff action, can their string environment variable be the comma-separated list of checks instead of the name of a file with the list of checks?

**Review by raghavkaul — CHANGES_REQUESTED** [review]
Thanks for working on this, some suggestions for code cleanup.

File: dependency-analysis/action.yaml
Comment (naveensrinivasan) [reply dependency-analysis/action.yaml:23]: I want to merge this in and associate it with the pinned SHA.

File: .github/workflows/publish-dependency-image.yml
Comment (naveensrinivasan) [reply .github/workflows/publish-dependency-image.yml:11]: We can do that. Probably after this PR gets merged in then, we can update it. I would like to get this merged so that we can beta-test it. I agree before going PROD we will update it.

File: dependency-analysis/main.go
Comment (naveensrinivasan) [reply dependency-analysis/main.go:305]: > would it make sense to move filename to a paremeter?
> That would centralize all the environment variable reading in `main()` and would be easier to test this function.

I can move the ENV variable to main. 
>also wonder if filename is the right approach? If the user is configuring the DependencyDiff action, can their string environment variable be the comma-separated list of checks instead of the name of a file with the list of checks?

The reason for the file is to make is cleaner to configure instead of passing `19` checks as an argument in the command line. 

Also when we have the policy it is also going to be a file and that was my thought process

File: dependency-analysis/main.go
Comment (naveensrinivasan) [reply dependency-analysis/main.go:95]: It will continue only when it is an error `GetScorecardResult`.

File: dependency-analysis/README.md
Comment (naveensrinivasan) [reply dependency-analysis/README.md:14]: Yes, I agree. That is for later.

File: dependency-analysis/main.go
Comment (naveensrinivasan) [reply dependency-analysis/main.go:276]: The go-github doesn't provide this API, and that was the reason for me to use this.

File: dependency-analysis/main.go
Comment (naveensrinivasan) [reply dependency-analysis/main.go:74]: We cannot pass the version as we have only commit SHA in Scorecard API.

File: dependency-analysis/main.go
Comment (naveensrinivasan) [reply dependency-analysis/main.go:118]: The`GITHUB_ACTOR` https://docs.github.com/en/actions/learn-github-actions/variables would be different if a human runs it, so we won't be updating the existing comment but will be creating new one.

Do you have other thoughts?

File: dependency-analysis/main.go
Comment (naveensrinivasan) [reply dependency-analysis/main.go:235]: For now, can we skip this please?

File: dependency-analysis/main.go
Comment (naveensrinivasan) [reply dependency-analysis/main.go:97]: Sorry, I don't get your point. I am not re-running any check other than filtering it here.

File: dependency-analysis/README.md
Comment (naveensrinivasan) [reply dependency-analysis/README.md]: 👍 , Fixed it. Thanks

File: dependency-analysis/README.md
Comment (naveensrinivasan) [reply dependency-analysis/README.md]: 👍

File: dependency-analysis/main.go
Comment (naveensrinivasan) [reply dependency-analysis/main.go]: Agreed, Thanks.

File: dependency-analysis/main.go
Comment (naveensrinivasan) [reply dependency-analysis/main.go]: 👍

File: dependency-analysis/main.go
Comment (naveensrinivasan) [reply dependency-analysis/main.go]: 👍

File: dependency-analysis/main.go
Comment (naveensrinivasan) [reply dependency-analysis/main.go:271]: Can we skip this for now?

File: dependency-analysis/main.go
Comment (naveensrinivasan) [reply dependency-analysis/main.go]: 👍

File: dependency-analysis/README.md
Comment (naveensrinivasan) [reply dependency-analysis/README.md]: 👍

File: dependency-analysis/main.go
Comment (naveensrinivasan) [reply dependency-analysis/main.go]: 👍

File: dependency-analysis/main.go
Comment (raghavkaul) [reply dependency-analysis/main.go]: I mean the dependency version - if we're running scorecard on the dependency, is it possible to indicate which version of the dependency we're scoring?

File: dependency-analysis/main.go
Comment (raghavkaul) [reply dependency-analysis/main.go:94]: Yes, but can we apply the filter function earlier so that we don't run scorecard checks that we'll filter out anyway?

File: dependency-analysis/main.go
Comment (raghavkaul) [reply dependency-analysis/main.go:110]: If we use a hardcoded username (like "scorecard-dependencydiff" for example) instead of `GITHUB_ACTOR`, it would avoid the issue where DependencyDiff attempts to post a comment with the name of the user who triggered the action.

File: dependency-analysis/main.go
Comment (naveensrinivasan) [reply dependency-analysis/main.go:94]: Again this is an API call to api.securityscorecards.dev, so it shouldn't be an issue as we are just fetching the results for the dependency.

File: dependency-analysis/main.go
Comment (naveensrinivasan) [reply dependency-analysis/main.go:110]: My logic searches for GitHub comments by the USER, and AFAIK, I cannot randomly assign a user that would create a comment with that user as a reference. So the comment would be created by the `GITHUB_ACTOR`. Unless you think it is possible?

File: dependency-analysis/main.go
Comment (raghavkaul) [reply dependency-analysis/main.go:94]: Got it, makes sense.

**Comment by azeemshaikh38** [comment]
Added a question/comment on this https://github.com/ossf/scorecard-action/issues/1070#issuecomment-1448404581.

**Review by azeemshaikh38 — CHANGES_REQUESTED** [review]
Question/comment on this design: https://github.com/ossf/scorecard-action/issues/1070#issuecomment-1448404581

File: dependency-analysis/main.go
Comment (raghavkaul) [reply dependency-analysis/main.go:110]: Ah I just found the docs that mention the difference between `actor` (which stays the same when your workflow is rerun manually) and `triggering_actor`. This is fine.

**Review by raghavkaul — APPROVED** [review]
Thanks, LGTM.

**Comment by naveensrinivasan** [comment]
> Question/comment on this design: [#1070 (comment)](https://github.com/ossf/scorecard-action/issues/1070#issuecomment-1448404581)

We are continuing the discussion on the ☝️ thread

**Comment by naveensrinivasan** [comment]
> Question/comment on this design: [#1070 (comment)](https://github.com/ossf/scorecard-action/issues/1070#issuecomment-1448404581)

A friendly ping

File: dependency-analysis/main.go
Comment (laurentsimon) [inline dependency-analysis/main.go:85]: Why are we treating Vulns as a special case? 

I worry our data is inaccurate, ie we're using a dependency HEAD whereas a user is installing a particular package version. There are other tools that will do a better job at this than Scorecard, like dependabot.

I'm thinking we may even remove the vuln results entirely because of how poor our data may be. Otherwise users will be confused and unhappy that we're claiming there are vulns in a package if there are none.

File: dependency-analysis/main.go
Comment (laurentsimon) [inline dependency-analysis/main.go:214]: maybe create a template file rather than hardcoded this layout in code. It' pretty hard to read otherwise

File: dependency-analysis/main.go
Comment (laurentsimon) [inline dependency-analysis/main.go:317]: Should we instead be returning just the aggregate score and link to deps.dev for better UX? Or at least provide a link to deps.dev for users to see more results?

File: dependency-analysis/main_test.go
Comment (laurentsimon) [inline dependency-analysis/main_test.go:101]: nit: use [go-cmp](https://github.com/google/go-cmp) package for deep comparison

File: dependency-analysis/README.md
Comment (naveensrinivasan) [reply dependency-analysis/README.md:9]: >are we showing the aggregate score in the summary for each result

We aren't showing aggregate score. Do you think we should show the aggregate score?

>Are we creating a new comment for each run?

We will not create results for reach run. Only if the user manually runs it only then we create another comment.

File: dependency-analysis/README.md
Comment (naveensrinivasan) [reply dependency-analysis/README.md:12]: > Is this really true? For go projects, unlikely. For other projects, is it? We don't need an exact dependency tag, so it should not be needed. Does GitHub API return dependency name / repo without enabling this feature?

I have tested for go projects and here is python package https://github.com/ossf-tests/vulpy/pull/2/files. For teh GH API to return results it needs this feature to be tuned on.

File: dependency-analysis/README.md
Comment (naveensrinivasan) [reply dependency-analysis/README.md:17]: We can give users the option to exclude certain repositories or checks for a particular repo, which we can make possible with the file option. 

This will also give us the ability to extend the feature in the future. That's the idea.

File: dependency-analysis/README.md
Comment (naveensrinivasan) [reply dependency-analysis/README.md:35]: I haven't tested with a private repo. I will test it and update it.

File: dependency-analysis/action.yaml
Comment (naveensrinivasan) [reply dependency-analysis/action.yaml:23]: Yes it is till we cut a release. 👍

File: dependency-analysis/README.md
Comment (naveensrinivasan) [reply dependency-analysis/README.md:42]: ${{ github.event.number }} isn't directly accessible within the GH action.

File: dependency-analysis/main.go
Comment (naveensrinivasan) [reply dependency-analysis/main.go:63]: This is GH Client access.
```
ts := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: token})
	tc := oauth2.NewClient(context.Background(), ts)
	client := github.NewClient(tc)
```	

I think from the code we should be able to make it being used in GH Client API. Unless you think it is still necessary.

File: dependency-analysis/main.go
Comment (naveensrinivasan) [reply dependency-analysis/main.go:85]: OK. I agree. I can remove the Vulns.

File: dependency-analysis/main.go
Comment (naveensrinivasan) [reply dependency-analysis/main.go:214]: I agree. I can do that.

File: dependency-analysis/main.go
Comment (naveensrinivasan) [reply dependency-analysis/main.go:317]: We can add the aggregate score but not every repo has the deps.dev and that is the reason not to include deps.dev.

File: dependency-analysis/README.md
Comment (laurentsimon) [reply dependency-analysis/README.md:17]: We should have a config file in this case then. Standard way to pass parameters to an Action is not env variables. It's inputs.

File: dependency-analysis/main.go
Comment (laurentsimon) [reply dependency-analysis/main.go:63]: auth token must bee needed to use the GITHUB_TOKEN. That makes sense now :-)

File: dependency-analysis/README.md
Comment (laurentsimon) [reply dependency-analysis/README.md:12]: Very interesting... and unexpected to me

**Review by azeemshaikh38 — APPROVED** [review]

File: dependency-analysis/README.md
Comment (laurentsimon) [reply dependency-analysis/README.md:9]: so we run it once for the first commit, and then only if the user asks to run?
Maybe add to the comment description to explain to users that they can re-run to update?

**Comment by spencerschrock** [comment]
Closing as this was implemented in a separate repo instead.
https://github.com/ossf/scorecard-dependencyanalysis

**Pull request closed without merging**
